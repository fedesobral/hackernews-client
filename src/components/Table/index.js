import React, { Component } from 'react';
import PropTypes from 'prop-types';
import classNames from 'classnames';
import { sortBy } from 'lodash';
import Button from '../Button'
import './index.css'

const SORTS = {
  NONE: list => list,
  TITLE: list => sortBy(list, 'title'),
  AUTHOR: list => sortBy(list, 'author'),
  COMMENTS: list => sortBy(list, 'num_comments'),
  POINTS: list => sortBy(list, 'points'),
}

const isSearched = (searchTerm) => (item) => {
  return !searchTerm || !item.title || item.title.toLowerCase().includes(searchTerm.toLowerCase());
};

export class Table extends Component {
  constructor(props){
    super(props);

    this.state = { 
      sortKey: 'NONE',
      isSortReverse: false
    };

    this.onSort = this.onSort.bind(this);
  }

  onSort(sortKey){
    this.setState((state, props) => ({ sortKey, isSortReverse: sortKey === state.sortKey && !state.isSortReverse  }));
  }

  render() {
    const { list, pattern, onDismiss } = this.props;
    const { sortKey, isSortReverse } = this.state;
    const sortedList = SORTS[sortKey](list.filter(isSearched(pattern)));
    const reverseSortedList = isSortReverse 
      ? sortedList.reverse()
      : sortedList;
    return (
      <div className='table'>
        <TableHeader 
          onSort={this.onSort}
          activeSortKey={sortKey}
          isSortReverse={isSortReverse}
        ></TableHeader>
        <div className='table-body'> 
        { 
          reverseSortedList.map(item => (
            <TableRow key={item.objectID} item={item} onDismiss={onDismiss} />) ) 
        }
        </div>
      </div>
    );
  }
}

Table.propTypes = {
  list: PropTypes.arrayOf(
      PropTypes.shape({
        objectID: PropTypes.string,
        url: PropTypes.string, 
        title: PropTypes.string, 
        author: PropTypes.string, 
        num_comments: PropTypes.number, 
        points: PropTypes.number
      })
    ).isRequired,
  pattern: PropTypes.string,
  onDismiss: PropTypes.func
}

const TableRow = ({item, onDismiss}) => {
    const {objectID, url, title, author, num_comments, points} = item;
    return (
      <div className='table-row'>
        <span className='midColumn'>
        <a href={url}>{title}</a>
        </span>
        <span className='midColumn'>{author}</span>
        <span className='smallColumn'>{num_comments}</span>
        <span className='smallColumn'>{points}</span>
        <span className='smallColumn'>
        <Button onClick={() => onDismiss(objectID)}
        className='button-inline' >
        Dismiss
        </Button> 
        </span>
      </div>
    );
}

const TableHeader = ({ onSort, activeSortKey, isSortReverse }) => {
  return (
    <div className='table-header'>
      <span className='midColumn'>
        <Sort 
          sortKey={'TITLE'}
          onSort={onSort}
          activeSortKey={activeSortKey}
          isSortReverse={isSortReverse}
         >Title</Sort>
      </span>
      <span className='midColumn'>
        <Sort 
          sortKey={'AUTHOR'}
          onSort={onSort}
          activeSortKey={activeSortKey}
          isSortReverse={isSortReverse}
         >Author</Sort>
      </span>
      <span className='smallColumn'>
        <Sort 
          sortKey={'COMMENTS'}
          onSort={onSort}
          activeSortKey={activeSortKey}
          isSortReverse={isSortReverse}
         >Comments</Sort>
      </span>
      <span className='smallColumn'>
        <Sort 
          sortKey={'POINTS'}
          onSort={onSort}
          activeSortKey={activeSortKey}
          isSortReverse={isSortReverse}
         >Points</Sort>
      </span>
      <span className='smallColumn'>
      <Button>Archive</Button>
      </span>
    </div>
    );
}

const Sort = ({ sortKey, onSort, activeSortKey, isSortReverse, children }) =>{
  const sortClass = classNames(
    'button-inline',
    {'button-active': sortKey === activeSortKey}
  );
  const sortOrderClass = classNames(
    'fa',
    {'fa-caret-up':  sortKey === activeSortKey && !isSortReverse},
    {'fa-caret-down':  sortKey === activeSortKey && isSortReverse},
  );
  return (
    <Button 
      onClick={() => onSort(sortKey) }
      className={sortClass}
    >
      {children}
      <span className={sortOrderClass}></span>
    </Button>
    );
}

TableRow.propTypes = {
  item: PropTypes.shape({
      objectID: PropTypes.string,
      url: PropTypes.string, 
      title: PropTypes.string, 
      author: PropTypes.string, 
      num_comments: PropTypes.number, 
      points: PropTypes.number
  }).isRequired,
  onDismiss: PropTypes.func
}

export default Table;
