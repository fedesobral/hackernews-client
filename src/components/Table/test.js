import React from 'react';
import ReactDOM from 'react-dom';
import { shallow, render, mount } from 'enzyme';
import { shallowToJson, mountToJson, renderToJson } from 'enzyme-to-json';
import Table from '../Table';

describe('Table', () => {

	const props = {
		list: [
			{ objectID: '1x', title: '1', author: '1', num_comments: 1, points: 1 },
			{ objectID: '2y', title: '2', author: '2', num_comments: 2, points: 2, url: 'www.google.com' },
		]
	}

	it('renders', () => {
	  const div = document.createElement('div');
	  ReactDOM.render(<Table { ...props } />, div);
	});

	test('snapshots', () => {
		const component = render(
			<Table { ...props } />
		);
		let tree = renderToJson(component);
		expect(tree).toMatchSnapshot();
	});

	it('show 2 items in list', () => {
		const element = render(
			<Table { ...props } />
		);

		expect(element.find('.table-row').length).toBe(2);
	});

	it('show correct number of items in list', () => {
		const element = mount(
			<Table { ...props } />
		);

		expect(element.find('.table-row').length).toBe(2);

		element.setProps({ 
			list: [
				{ objectID: '1x', title: '1', author: '1', num_comments: 1, points: 1 },
				{ objectID: '2y', title: '2', author: '2', num_comments: 2, points: 2, url: 'www.google.com' },
				{ objectID: '3z', title: '3', author: '3', num_comments: 3, points: 3, url: 'www.google.com' },
			]
		});
		expect(element.find('.table-row').length).toBe(3);
	});

	it('sorted', () => {
		const sortProps = { 
			list: props.list,
			sortKey: 'TITLE', 
			activeSortKey: 'TITLE', 
			isSortReverse: true
		};

		const element = mount(
			<Table { ...sortProps } />
		);

		expect(mountToJson(element)).toMatchSnapshot();
	});
});