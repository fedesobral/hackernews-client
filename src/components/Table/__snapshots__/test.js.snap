exports[`Table snapshots 1`] = `
<div
  class="table">
  <div
    class="table-header">
    <span
      class="midColumn">
      <button
        class="button-inline"
        type="button">
        Title
        <span
          class="fa" />
      </button>
    </span>
    <span
      class="midColumn">
      <button
        class="button-inline"
        type="button">
        Author
        <span
          class="fa" />
      </button>
    </span>
    <span
      class="smallColumn">
      <button
        class="button-inline"
        type="button">
        Comments
        <span
          class="fa" />
      </button>
    </span>
    <span
      class="smallColumn">
      <button
        class="button-inline"
        type="button">
        Points
        <span
          class="fa" />
      </button>
    </span>
    <span
      class="smallColumn">
      <button
        class=""
        type="button">
        Archive
      </button>
    </span>
  </div>
  <div
    class="table-body">
    <div
      class="table-row">
      <span
        class="midColumn">
        <a>
          1
        </a>
      </span>
      <span
        class="midColumn">
        1
      </span>
      <span
        class="smallColumn">
        1
      </span>
      <span
        class="smallColumn">
        1
      </span>
      <span
        class="smallColumn">
        <button
          class="button-inline"
          type="button">
          Dismiss
        </button>
      </span>
    </div>
    <div
      class="table-row">
      <span
        class="midColumn">
        <a
          href="www.google.com">
          2
        </a>
      </span>
      <span
        class="midColumn">
        2
      </span>
      <span
        class="smallColumn">
        2
      </span>
      <span
        class="smallColumn">
        2
      </span>
      <span
        class="smallColumn">
        <button
          class="button-inline"
          type="button">
          Dismiss
        </button>
      </span>
    </div>
  </div>
</div>
`;

exports[`Table sorted 1`] = `
<Table
  activeSortKey="TITLE"
  isSortReverse={true}
  list={
    Array [
      Object {
        "author": "1",
        "num_comments": 1,
        "objectID": "1x",
        "points": 1,
        "title": "1",
      },
      Object {
        "author": "2",
        "num_comments": 2,
        "objectID": "2y",
        "points": 2,
        "title": "2",
        "url": "www.google.com",
      },
    ]
  }
  sortKey="TITLE">
  <div
    className="table">
    <TableHeader
      activeSortKey="NONE"
      isSortReverse={false}
      onSort={[Function]}>
      <div
        className="table-header">
        <span
          className="midColumn">
          <Sort
            activeSortKey="NONE"
            isSortReverse={false}
            onSort={[Function]}
            sortKey="TITLE">
            <Button
              className="button-inline"
              onClick={[Function]}>
              <button
                className="button-inline"
                onClick={[Function]}
                type="button">
                Title
                <span
                  className="fa" />
              </button>
            </Button>
          </Sort>
        </span>
        <span
          className="midColumn">
          <Sort
            activeSortKey="NONE"
            isSortReverse={false}
            onSort={[Function]}
            sortKey="AUTHOR">
            <Button
              className="button-inline"
              onClick={[Function]}>
              <button
                className="button-inline"
                onClick={[Function]}
                type="button">
                Author
                <span
                  className="fa" />
              </button>
            </Button>
          </Sort>
        </span>
        <span
          className="smallColumn">
          <Sort
            activeSortKey="NONE"
            isSortReverse={false}
            onSort={[Function]}
            sortKey="COMMENTS">
            <Button
              className="button-inline"
              onClick={[Function]}>
              <button
                className="button-inline"
                onClick={[Function]}
                type="button">
                Comments
                <span
                  className="fa" />
              </button>
            </Button>
          </Sort>
        </span>
        <span
          className="smallColumn">
          <Sort
            activeSortKey="NONE"
            isSortReverse={false}
            onSort={[Function]}
            sortKey="POINTS">
            <Button
              className="button-inline"
              onClick={[Function]}>
              <button
                className="button-inline"
                onClick={[Function]}
                type="button">
                Points
                <span
                  className="fa" />
              </button>
            </Button>
          </Sort>
        </span>
        <span
          className="smallColumn">
          <Button
            className="">
            <button
              className=""
              type="button">
              Archive
            </button>
          </Button>
        </span>
      </div>
    </TableHeader>
    <div
      className="table-body">
      <TableRow
        item={
          Object {
            "author": "1",
            "num_comments": 1,
            "objectID": "1x",
            "points": 1,
            "title": "1",
          }
        }>
        <div
          className="table-row">
          <span
            className="midColumn">
            <a>
              1
            </a>
          </span>
          <span
            className="midColumn">
            1
          </span>
          <span
            className="smallColumn">
            1
          </span>
          <span
            className="smallColumn">
            1
          </span>
          <span
            className="smallColumn">
            <Button
              className="button-inline"
              onClick={[Function]}>
              <button
                className="button-inline"
                onClick={[Function]}
                type="button">
                Dismiss
              </button>
            </Button>
          </span>
        </div>
      </TableRow>
      <TableRow
        item={
          Object {
            "author": "2",
            "num_comments": 2,
            "objectID": "2y",
            "points": 2,
            "title": "2",
            "url": "www.google.com",
          }
        }>
        <div
          className="table-row">
          <span
            className="midColumn">
            <a
              href="www.google.com">
              2
            </a>
          </span>
          <span
            className="midColumn">
            2
          </span>
          <span
            className="smallColumn">
            2
          </span>
          <span
            className="smallColumn">
            2
          </span>
          <span
            className="smallColumn">
            <Button
              className="button-inline"
              onClick={[Function]}>
              <button
                className="button-inline"
                onClick={[Function]}
                type="button">
                Dismiss
              </button>
            </Button>
          </span>
        </div>
      </TableRow>
    </div>
  </div>
</Table>
`;
